File: src/main/java/ru/citeck/idea/templates/project/CiteckModuleBuilder.java
package ru.citeck.idea.templates.project;

import com.intellij.ide.util.projectWizard.ModuleWizardStep;
import com.intellij.ide.util.projectWizard.WizardContext;
import com.intellij.openapi.Disposable;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.DumbAwareRunnable;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.roots.ui.configuration.ModulesProvider;
import com.intellij.openapi.util.io.FileUtil;
import com.intellij.openapi.vfs.VirtualFile;
import icons.Icons;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.idea.maven.execution.MavenRunner;
import org.jetbrains.idea.maven.execution.MavenRunnerParameters;
import org.jetbrains.idea.maven.model.MavenConstants;
import org.jetbrains.idea.maven.utils.MavenUtil;
import org.jetbrains.idea.maven.wizards.AbstractMavenModuleBuilder;
import org.jetbrains.idea.maven.wizards.MavenOpenProjectProvider;
import ru.citeck.idea.utils.CiteckVirtualFileUtils;

import javax.swing.*;
import java.io.File;
import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class CiteckModuleBuilder extends AbstractMavenModuleBuilder {

    public static final String MAVEN_PLUGIN_GENERATE_ARCHETYPE = "org.apache.maven.plugins:maven-archetype-plugin:2.4:generate";
    public static final String REPOSITORY = "https://nexus.citeck.ru/repository/maven-public/";

    public CiteckModuleBuilder() {
    }

    @Override
    public @Nullable ModuleWizardStep getCustomOptionsStep(WizardContext context, Disposable parentDisposable) {
        return new EcosModuleWizardStep(this);
    }

    @Override
    public String getPresentableName() {
        return "Citeck";
    }

    @Override
    public String getDescription() {
        return "Create <b>Citeck</b> project";
    }

    @Override
    public Icon getNodeIcon() {
        return Icons.CiteckLogo;
    }

    @Override
    protected void setupModule(Module module) {
        Project project = module.getProject();
        MavenUtil.runWhenInitialized(project, (DumbAwareRunnable) () ->
            ApplicationManager.getApplication().invokeLater(() -> {
                try {
                    generateMavenArchetype(module);
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }));
    }

    private void generateMavenArchetype(Module module) throws IOException {

        Project project = module.getProject();
        VirtualFile moduleDir = CiteckVirtualFileUtils.getFileByPath(getModuleFileDirectory());

        File tmpDir = FileUtil.createTempDirectory("archetype", "tmp");
        tmpDir.deleteOnExit();

        MavenRunnerParameters mavenParams = new MavenRunnerParameters(
            false, tmpDir.getPath(), (String) null,
            List.of(MAVEN_PLUGIN_GENERATE_ARCHETYPE),
            Collections.emptyList());

        MavenRunner runner = MavenRunner.getInstance(project);
        Map<String, String> mavenProperties = runner.getSettings().getMavenProperties();
        mavenProperties.putAll(this.getPropertiesToCreateByArtifact());
        mavenProperties.put("archetypeCatalog", REPOSITORY);
        mavenProperties.put("interactiveMode", "false");

        runner.run(mavenParams, runner.getState().clone(), () -> {
            try {
                String artifactId = getPropertiesToCreateByArtifact().get("artifactId");
                FileUtil.copyDir(new File(tmpDir, artifactId), new File(moduleDir.getPath()));
                FileUtil.delete(tmpDir);
                moduleDir.refresh(false, false);
                Optional
                    .ofNullable(moduleDir.findChild(MavenConstants.POM_XML))
                    .ifPresent(pom -> new MavenOpenProjectProvider().linkToExistingProject(pom, project));
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        });
    }

    @Override
    public ModuleWizardStep[] createWizardSteps(@NotNull WizardContext wizardContext,
                                                @NotNull ModulesProvider modulesProvider) {
        return new ModuleWizardStep[0];
    }
}

File: src/main/kotlin/ru/citeck/idea/newproject/CiteckModuleBuilder.kt
package ru.citeck.idea.newproject

import com.intellij.ide.util.projectWizard.ModuleWizardStep
import com.intellij.ide.util.projectWizard.WizardContext
import com.intellij.openapi.Disposable
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.module.Module
import com.intellij.openapi.project.DumbAwareRunnable
import com.intellij.openapi.roots.ui.configuration.ModulesProvider
import com.intellij.openapi.util.io.FileUtil
import com.intellij.openapi.vfs.VirtualFile
import icons.Icons
import org.jetbrains.idea.maven.execution.MavenRunner
import org.jetbrains.idea.maven.execution.MavenRunnerParameters
import org.jetbrains.idea.maven.model.MavenConstants
import org.jetbrains.idea.maven.utils.MavenUtil
import org.jetbrains.idea.maven.wizards.AbstractMavenModuleBuilder
import org.jetbrains.idea.maven.wizards.MavenOpenProjectProvider
import ru.citeck.idea.utils.CiteckVirtualFileUtils.getFileByPath
import java.io.File
import java.io.IOException
import java.util.*
import javax.swing.Icon
import kotlin.collections.emptyList
import kotlin.collections.set

class CiteckModuleBuilder : AbstractMavenModuleBuilder() {

    companion object {
        const val MAVEN_PLUGIN_GENERATE_ARCHETYPE: String =
            "org.apache.maven.plugins:maven-archetype-plugin:2.4:generate"
        const val REPOSITORY: String = "https://nexus.citeck.ru/repository/maven-public/"
    }

    override fun getCustomOptionsStep(context: WizardContext, parentDisposable: Disposable): ModuleWizardStep {
        return CiteckModuleWizardStep(this)
    }

    override fun getPresentableName(): String {
        return "Citeck"
    }

    override fun getDescription(): String {
        return "Create <b>Citeck</b> project"
    }

    override fun getNodeIcon(): Icon {
        return Icons.CiteckLogo
    }

    override fun setupModule(module: Module) {
        val project = module.project
        MavenUtil.runWhenInitialized(
            project,
            DumbAwareRunnable {
                ApplicationManager.getApplication().invokeLater {
                    try {
                        generateMavenArchetype(module)
                    } catch (e: IOException) {
                        throw RuntimeException(e)
                    }
                }
            })
    }

    private fun generateMavenArchetype(module: Module) {
        val project = module.project
        val moduleDir = getFileByPath(
            moduleFileDirectory!!
        )

        val tmpDir = FileUtil.createTempDirectory("archetype", "tmp")
        tmpDir.deleteOnExit()

        val mavenParams = MavenRunnerParameters(
            false, tmpDir.path, null as String?,
            listOf(MAVEN_PLUGIN_GENERATE_ARCHETYPE),
            emptyList()
        )

        val runner = MavenRunner.getInstance(project)
        val mavenProperties = runner.settings.mavenProperties
        mavenProperties.putAll(this.propertiesToCreateByArtifact)
        mavenProperties["archetypeCatalog"] = REPOSITORY
        mavenProperties["interactiveMode"] = "false"

        runner.run(mavenParams, runner.state.clone()) {
            try {
                val artifactId = propertiesToCreateByArtifact["artifactId"] ?: "unknown"
                FileUtil.copyDir(
                    File(tmpDir, artifactId),
                    File(moduleDir!!.path)
                )
                FileUtil.delete(tmpDir)
                moduleDir.refresh(false, false)
                Optional
                    .ofNullable(moduleDir.findChild(MavenConstants.POM_XML))
                    .ifPresent { pom: VirtualFile? ->
                        MavenOpenProjectProvider().linkToExistingProject(
                            pom!!, project
                        )
                    }
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
    }

    override fun createWizardSteps(
        wizardContext: WizardContext,
        modulesProvider: ModulesProvider
    ): Array<ModuleWizardStep> {
        return emptyArray()
    }
}
